### ISSUE TITLE
Allow management of Kubelet configuration in CAPI

### ISSUE THREAD

**Issue by fabriziopandini** [kubernetes-sigs/cluster-api#4464]
**User Story**

As a CAPI user I would like to manage the KubeletConfiguration in a declarative way

**Detailed Description**

This is a follow up of https://github.com/kubernetes-sigs/cluster-api/issues/1584.

As of today when initialising a cluster with CABPK, a default KubeletConfiguration is used and stored into `kubelet-config-XX` ConfigMap gets created, and the same config map is then carried away during upgrades.

The user can edit the `kubelet-config-XX` ConfigMap between upgrades, but this approach is not declarative and does not provide a good abstraction over this configuration detail in CAPI.

This issue is about discussing following options:
- Allow the user to specify the KubeletConfiguration to be used for creating the cluster in KCP.
- Use the KubeletConfiguration specified in KCP as a authoritative source of truth for the `kubelet-config-XX` ConfigMap (and thus reconcile it, preventing direct modification/drift from the authoritative source)
- Allow the user to edit the KubeletConfiguration cluster in KCP, and use the new version for joining new nodes.

A few consideration should be addressed:
- How to model the KubeletConfiguration into KCP or into the KubeadmConfig, given that 
  - The KubeletConfiguration API is still evolving, and embedding a not stable API ultimately results in an unstable API surface for the users (it is also hard to maintain given that introduces hard dependency between two projects with different lifecycles)
  - The KubeletConfiguration makes sense at cluster level only because as of today it is a global configuration in kubeadm (Nb. node specific configuration are already supported using ExtraArgs into the NodeRegistration object)
- How to manage following scenarios
  - Adding a user managed KubeletConfiguration to a cluster using a default one (seamless upgrade/adoption)
  - Changing a user managed KubeletConfiguration (Should this trigger rollout in KCP? what about other nodes in the cluster)
  - Removing a user managed KubeletConfiguration to a cluster using a user managed one (is this allowed? if yes, should we rollback to the default one or keep the last user provided one and carry over across versions/without reconciling?)
  
**Anything else you would like to add:**

/kind feature

**Comment by sbueringer** [comment]
If possible, I think it would be really nice to be able to use different KubeletConfiguration for control plane and worker nodes.

Our experience is that depending on the node size the eviction thresholds should be configured differently. E.g. control plane nodes with 4 cpu / 16 GI require different thresholds then worker nodes with 16 cpu / 16 Gi RAM.

On our internal platform we solved this by deploying KubeletConfigurations and assigning them to nodes after the `kubeadm join` is done.

**Comment by detiber** [comment]
I agree with @sbueringer, if we allow this we should allow for a configuration on a more granular level than an entire cluster. I would argue probably allow for configuration at the KCP, MachinePool, and MachineDeployment/MachineSet level, that way one could get as granular as they wanted.

The one thing that I do worry about with this, though is that currently doesn't kubeadm manage the KubeletConfiguration at the version level? Are we going to end up having to handle any type of conversions or other automated mutations for users if they specify a KubeletConfiguration? Or are we going to get into odd situations where specifying the KubeletConfiguration will interfere with upgrade handling?

**Comment by sbueringer** [comment]
I'm not sure about the impact of updates. Just some details about our current solution:
* per Node
  * create server in IaaS
  * kubeadm init / join
  * deploy KubeletConfigurationConfigMap
  * set Node KubeletConfig on the Node (spec/configSource/ConfigMap/*)
  * wait until KubeletConfig active (i.e. Node.Status.Config.Active.ConfigMap.{Name,UID,ResourceVersion} matches to the CM deployed before)

We don't execute commands like `kubeadm update`, but afaik nor does ClusterAPI. We create the KubeletConfiguration ConfigMaps with distinct names to the one used by `kubeadm`.  We explicitly wait until kubeadm is done before changing the KubeletConfiguration. Up until now we didn't have any problems with this approach, but if we want to use this approach we should definitely ensure that kubeadm does not overwrite the active KubeletConfiguration once we changed it.

**Comment by neolit123** [comment]
i think the best option is to just allow passing KubeletConfiguration to the `kubeadm join` command and apply some sort of a marker (e.g. `# some yaml comment`) to this kubelet config.yaml is no longer managed by kubeadm:
https://github.com/kubernetes/kubeadm/issues/1682

as far as the kubelet CM naming goes, we really want kubeadm to stop versioning the CM name:
https://github.com/kubernetes/kubeadm/issues/1582
but this would be a breaking change to a lot of users and we need to deprecate it and manage the two differently named CMs in parallel for 3+ releases.

**Comment by neolit123** [comment]
but to share, @fabriziopandini and me had a lengthy discussion about the state of kubeletconfiguration vs kubelet flags vs instance specific config vs component config upgrade and overall this space is quite messy and problematic.

@fabriziopandini preferred that we continue using the pattern on passing a single config to kubeadm and then users could apply node specific overrides with kubeletExtraArgs.

of course, this has the risk that if the kubelet suddenly removes a number of its deprecated flags users of CAPI will be in trouble.

**Comment by sbueringer** [comment]
@neolit123 
So from a CAPI perspective:
* deploy KubeletConfiguration ConfigMap
* pass the name of the KubeletConfiguration to kubeadm, kubeadm then ensure that the KubeletConfiguration ConfigMap is used and we don't have to.

Possible problem:
* we can only change the KubeletConfiguration when creating new nodes, we cannot dynamically change it (but maybe that wasn't the plan anyway)

> of course, this has the risk that if the kubelet suddenly removes a number of its deprecated flags, users of CAPI will be in trouble.
> (same goes for kubeadm users or users of other tools).

That was also one of my concerns. `kubeletExtraArgs` only work as long as we can configure all relevant configuration there. So not sure if that's a good long-term solution.

**Comment by enxebre** [comment]
Relates to https://github.com/kubernetes-sigs/cluster-api/issues/4444

**Comment by fabriziopandini** [comment]
This issue is about allowing a declarative approach around KubeletConfiguration as it is supported in kubeadm today (cluster level configuration).

Instance/group of instance specific KubeletConfigurations is something useful, but the change should be implemented in kubeadm first, and only after in CAPI. However as @detiber and @neolit123 was pointing out:

- As of today there is not automatic conversion mechanism for the KubeletConfiguration
- the state of kubeletconfiguration vs kubelet flags vs instance specific config vs component config upgrade and overall this space is quite messy and problematic.

Given that, we should be careful on changing the status quo in kubeadm unless SIG-node/SIG-architecture clarifies the roadmap in this area. 
FYI with the help of some colleagues I'm trying to raise the point in the SIG-node agenda, but any help will be appreciated.

TL;DR; this issue focuses on what is supported today by kubeadm, which is already an improvement on the current status.
more invasive steps requires to act somewhere else and some clarity at kubernetes level

**Comment by sbueringer** [comment]
fyi. If I understood it correctly it should be possible to customize/patch the KubeletConfiguration with kubeadm/Kubernetes v1.25:
* https://github.com/kubernetes/kubeadm/issues/1682
* https://github.com/kubernetes/kubernetes/pull/110405

**Comment by fabriziopandini** [comment]
/close

Given that the future of component config is not clear I would avoid extending its usage in CAPI. Let's collect use cases where current approach does not work before committing to a way forward
